### ISSUE TITLE
Add client-side retrieval of UI search results

### ISSUE THREAD

**Issue by mikkonie** [bihealth/sodar-core#22]
**Update 2026-03:** I'm reviving this old issue and cleaning up old comments.

AT the moment, UI search and results displaying is performed 100% on the server side. This does not work great for all real-life use cases, as SODAR Core can't control the performance or waiting time needed for different apps to produce results.

Instead, we should have client-side JQuery perform plugin-specific queries for apps and display results for each app as they arrive.

Some considerations:

- We should maintain a specific ordering of the apps/plugins for each query, regardless of if and when they return results.
- We should clearly display a "no results" or "error querying for results" element for each queried plugin, maintaining the order mentioned in the previous point.
- DataTables needs to be set up for each plugin's result set, as it is in the current implementation.
- Breaking changes are needed for apps' `search()` return data (see below)

Problems with current method of app search results:

- Currently, apps are free to determine rendering the results table however they want
- With client-side retrieval, we can't do this anymore
    + ..unless we have the search API return HTML, which is obviously ill-advised

Solution:
 
- We meed to declare a standardized return data format for `search()`
    * Something to allow us to build the DataTable client-side
    * ..while supporting as many of the custom table rendering things currently done in the app result templates as possible
- This also needs to be nicely serializable into Ajax view return data for the client-side JQuery
- The basic sctructure would be something like:
    * List of table columns with column names and potential attributes regarding e.g. formatting
    * List of rows of data for the results table
- **NOTE:** One plugin may return multiple results tables depending on search type!
    * See e.g. samplesheets app in SODAR
    * To keep it simple, let's query for all results from the plugin in one request and then display 1-N tables depending on what the plugin returns

### Update ProjectSearchResultsView for Client-Side Search

- Add dict of app plugin search types to be searched in context data
    * Return list of `ProjectAppPlugin`:s with `search_enable=True`
    * Make sure to order this alphabetically by plugin name
    * For each plugin, include `plugin_name`, `plugin_title` and `icon`
- Disable server-side search results retrieval in the view
    * Don't delete the code yet, we'll have use for elsewhere :)
- (This might be a good time to update tests to remove expected server-side search results)

### Update search API and Implementations

- Update expected return data from project app plugin `search()` methods
    * `search()` should return 1-N result tables
        + Column & formatting info
        + Rows with the actual results
    * Require `search()` to provide results in a specific Python object format
        + Ajax view will serialize this to JSON for the client
    * **TBD:** Specific format
        + Table metadata (title + anything else?)
        + List of column metadata (title, possible formatting such as e.g. alignment)
        + List of lists for table rows
        + Cell object for each row cell (text, possible url, optional additional things such as icons?)
    * Let's start with as simple model as possible, then iterate if we figure out special cases
- Update `search()` implementations of existing apps to match the new design
- Remove now-unused search templates and plugin `search_template` values from apps
- Remove `ProjectAppPlugin.search_template`
- (Here it's a good idea to update tests again to match expectations)

### Create Ajax View

- Create Ajax view for querying for search results
- Queries should search by plugin
- Internally, this should incorporate the server side search code we had before, adapted for new changes of course
- Serialize `search()` result objects into JSON
- Return list of 0-N tables with data

### Add Client-Side Search

- In `search_results.html`, initially display a waiting spinner
- For each plugin in list, call Ajax view to return plugin-specific search results
    * I think we can just do fetch requests in parallel for all apps
- Populate tables once results are returned
    * Initialize DataTable etc
    * Maintain ordering of plugins!
- If no results are returned from a specific plugin, display a card with a "no results found" element

### Update Higlighting

- Refactor highlighting to work fully in JQuery
- Once all results have been returned, then go through results and highlight
- Should be considered "low priority" from the perspective of search execution
   
### Test

- Update `SearchResultsView` tests
- Update all `search()` implementation tests
- Add tests for new AjaxView
- Add UI tests for client-side retrieval

### Document

- Document requirements for `search()` results

## Tasks

- [ ] Update `ProjectSearchResultsView` logic
- [ ] Update search API
- [ ] Create Ajax view
- [ ] Add client-side search implementation
- [ ] Update highlighting
- [ ] Test
- [ ] Document

**Comment by mikkonie** [comment]
Old comments deleted as they were no longer informative. I'll be writing a spec in the original issue description.

**Comment by mikkonie** [comment]
Done by @fmarotta, merged.
